### PR TITLE
Color and font information for chars, words and boxes

### PR DESCRIPTION
**Pull request by kreuzberger**
The PR is for discussion of issue #38. The non-stroking color information is stored if all characters in a word have same non-stroking color.

### CODE REVIEW

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [inline libpdf/models/horizontal_box.py]: this would also run in case `x.ncolor` is None, right?

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [inline libpdf/models/horizontal_box.py]: The term `array` is already [taken](https://docs.python.org/3/library/array.html) in the Python world. What we have here is a list.

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [inline libpdf/models/horizontal_box.py]: This should be a recognizable type hint, so `Tuple[float, float, float] | None` or `Optional[Tuple[float, float, float]]` would be better. The information RGB should go to the line above.
(Haven't tested my type hint though)

**Review by ubmarco — CHANGES_REQUESTED** [review]
Thanks for the PR.
Besides my 2 comments, another question: I think it makes sense to extend the ncolor attribute to also show up on `HorizontalLine` and `HorizontalBox` in case all words/lines have the same ncolor value.

File: libpdf/models/horizontal_box.py
Comment (kreuzberger) [reply libpdf/models/horizontal_box.py]: currently the value taken is from PDF as rgb as floats. Should the value stored be converted to some other type (e.g. tuple of int in range [0,255] or to hex value? This could make it easier for comparisons.

Comment (ubmarco) [reply libpdf/models/horizontal_box.py]: For the number format I have no strong preference. Off these options I'd go with the PDF standard / floats. [Matplotlib](https://matplotlib.org/stable/users/explain/colors/colors.html) also uses it as numeric input.
Converting the colors by any calling library should be easy. Floats could have a higher accuracy than the 8 bits we get from 0.255 for each color. I guess there are people that go crazy about color depth and accuracy.

**Comment by ubmarco** [comment]
Can we try to merge https://github.com/useblocks/libpdf/pull/36 first to get the linting insights? It's almost done.

**Comment by ubmarco** [comment]
I merged PR https://github.com/useblocks/libpdf/pull/36, please add modified files to the tox.ini line that specifies the `ruff` paths, currently `poetry run ruff tests/conftest.py tests/test_rects.py`

**Comment by kreuzberger** [comment]
> I merged PR #36, please add modified files to the tox.ini line that specifies the `ruff` paths, currently `poetry run ruff tests/conftest.py tests/test_rects.py`

How can i use different python versions with tox to check? Currently i have 3.10 installed in my system and 3.11 in a local venv. How can is use 3.12 (required for format/lint) in a special local tox environment?
This would help me not to localy modify tox.ini all the time. or could the format/lint be more "Opened" for other python versions?

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [inline libpdf/models/horizontal_box.py]: This logic *should* not break, but I'm not sure about that. I've seen fonts that are exposed like this
image
Again I think this logic can be done by an end user.

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [inline libpdf/models/horizontal_box.py]: I think that logic can break easily or at least produce unexpected results. There is no real standard in font names, I think exposing the fontname is good enough. The bold logic can be determined by the end user (or us in case we write test cases on known PDFs). See here:
image
image

File: tests/test_word_colors.py
Comment (ubmarco) [inline tests/test_word_colors.py]:
```suggestion
                elif "italic text format" in content.textbox.text:
```

File: tests/test_word_colors.py
Comment (ubmarco) [inline tests/test_word_colors.py]:
```suggestion
                    if word.text == "italic":
```

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [reply libpdf/models/horizontal_box.py]: I see you went with float.

**Review by ubmarco — CHANGES_REQUESTED** [review]
Thx for the fixes & some more comments on the new changes.
I had a bad feeling putting the linting fixes of our code onto you, so I fixed all issues in utils.py and horizontal_box.py in https://github.com/useblocks/libpdf/pull/40. Will try to merge this tomorrow after a review. Please rebase your work onto this.

I use https://github.com/pyenv/pyenv to install any Python version I need.
If you cannot do that, just change the `basepython` in tox.ini to anything you have from 3.8 to 3.12. It _should_ produce the same output. The CI will tell us whether that assumption is valid.

File: libpdf/models/horizontal_box.py
Comment (kreuzberger) [reply libpdf/models/horizontal_box.py]: Ok, your are the PDF Expert :smile: . Exposing the fontname should follow the same strategy like for ncolor? store fontname on each char, return fontname for word if all chars in the word have same font and so on.

File: tests/test_word_colors.py
Comment (kreuzberger) [reply tests/test_word_colors.py]: Ok, your are the pdf expert :smile: . Exposing the fontname should follow the same strategy like for ncolor? store fontname on each char, return fontname for word if all chars in the word have same font and so on.

File: libpdf/models/horizontal_box.py
Comment (ubmarco) [reply libpdf/models/horizontal_box.py]: I just checked fonts installed in my system. And as PDF can embed fonts, my logical conclusion is that the logic may fail. But having the font name is a really good idea. Also font size would be good. Yes, I'd propagate it as well as you did for ncolor.

File: tests/test_word_colors.py
Comment (ubmarco) [reply tests/test_word_colors.py]: Haven't checked that, but I guess the `italc` typo is also in the PDF, right? If we remove the italic handling, this test code is to be removed anyway.

**Comment by ubmarco** [comment]
I just merged https://github.com/useblocks/libpdf/pull/40, please rebase your change on top of it. There might some merge conflicts. Btw, if you create your fork under your personal account, you can allow me to push to your PR branch (flag `Allow edits by maintainers`).

**Comment by kreuzberger** [comment]
> I just merged #40, please rebase your change on top of it. There might some merge conflicts. Btw, if you create your fork under your personal account, you can allow me to push to your PR branch (flag `Allow edits by maintainers`).

Hi! I merged everything some weeks ago, but the PR is still open. Is there something missing? Or Just no time left :smile:

**Review by ubmarco — CHANGES_REQUESTED** [review]
There is just one typo in the test case (`italc`). If that is fixed let's merge the PR!
Thanks for your work on this.

File: tests/test_word_colors.py
Comment (kreuzberger) [reply tests/test_word_colors.py]: Ups. fixed the typo, but not closed this conversation. But was good, line 138 also contained the typo :smile:

**Review by ubmarco — APPROVED** [review]
Awesome work, thanks for your contribution!

**Pull request merged**
